Source/WebCore/page/Page.cpp

```diff
@@ -791,17 +793,39 @@ void Page::setMainFrame(Ref<Frame>&& frame)
     m_mainFrame = WTFMove(frame);
 }
 
-void Page::setMainFrameURL(const URL& url)
+void Page::setMainFrameURLInternal(const URL& url)
 {
     m_mainFrameURL = url;
     m_mainFrameOrigin = SecurityOrigin::create(url);
 }
 
+void Page::setMainFrameURL(const URL& url)
+{
+    setMainFrameURLInternal(url);
```

Comment:
These could probably all have an early return if the thing your setting is equal to the thing that is already stored.